lib/authorizer/oauth2.js

```diff
@@ -58,6 +59,34 @@ module.exports = {
      * @param {AuthHandlerInterface~authPreHookCallback} done -
      */
     pre: function (auth, done) {
+        const authId = auth.get('authId'),
+            refreshTokenHelper = auth.get('refreshTokenHelper');
+
+        // Go into the refresh token flow ONLY IF we have all necessary information present
+        if (authId && refreshTokenHelper && _.isFunction(refreshTokenHelper.refreshToken)) {
+            const refreshTokenPromise = refreshTokenHelper.refreshToken(authId);
+
+            return Promise.race([
+                refreshTokenPromise,
+                new Promise((resolve) => {
+                    // We add a timeout to the refresh token promise to ensure that the request is not
+                    // hung up on the refresh token call. If the refresh token call takes more than
+                    // REFRESH_TOKEN_TIMEOUT_MS, we abandon and move forward with the execution.
+                    setTimeout(() => {
+                        return resolve();
```

Comment:
Why resolve? Wouldn't this cause the token to be reset with `undefined`?

Reply:
Nope, we only update the token if there's something present. Else we proceed as we would normally before.

Reply:
Wouldn't it make more sense to reject here since this is an error (request timed out)? Also, how are we communicating this to the client? Should the timeout be implemented in the handler?

Reply:
Was wondering if we should handle this in client or runtime, I think it's safer to have a timeout in runtime, if we add refresh support to monitors/newman. We would need them to implement a timeout as well, better to handle it here.

No communication to client right now. Wasn't sure if we should add an event for this case.